.dotnet/src/Custom/Internal/ClientPipelineExtensions.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.ClientModel;
+using System.ClientModel.Primitives;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenAI;
+
+internal static partial class ClientPipelineExtensions
+{
+    // CUSTOM:
+    // - Supplemented exception body with deserialized OpenAI error details
+
+    public static async ValueTask<PipelineResponse> ProcessMessageAsync(
+        this ClientPipeline pipeline,
+        PipelineMessage message,
+        RequestOptions options)
+    {
+        await pipeline.SendAsync(message).ConfigureAwait(false);
+
+        if (message.Response == null)
+        {
+            throw new InvalidOperationException("Failed to receive Result.");
+        }
+
+        if (message.Response.IsError && (options?.ErrorOptions & ClientErrorBehaviors.NoThrow) != ClientErrorBehaviors.NoThrow)
+        {
+            throw await TryBufferResponseAndCreateErrorAsync(message.Response).ConfigureAwait(false) switch
+            {
+                string errorMessage when !string.IsNullOrEmpty(errorMessage)
+                    => new ClientResultException(errorMessage, message.Response),
+                _ => new ClientResultException(message.Response),
+            };
+        }
+
+        return message.Response;
+    }
+        
+    public static PipelineResponse ProcessMessage(
+        this ClientPipeline pipeline,
+        PipelineMessage message,
+        RequestOptions options)
+    {
+        pipeline.Send(message);
+
+        if (message.Response.IsError && (options?.ErrorOptions & ClientErrorBehaviors.NoThrow) != ClientErrorBehaviors.NoThrow)
+        {
+            throw TryBufferResponseAndCreateError(message.Response) switch
+            {
+                string errorMessage when !string.IsNullOrEmpty(errorMessage)
+                    => new ClientResultException(errorMessage, message.Response),
+                _ => new ClientResultException(message.Response),
+            };
+        }
+
+        return message.Response;
+    }
+
+    private static string TryBufferResponseAndCreateError(PipelineResponse response)
+    {
+        response.BufferContent();
+        return TryCreateErrorMessageFromResponse(response);
+    }
+
+    private static async Task<string> TryBufferResponseAndCreateErrorAsync(PipelineResponse response)
+    {
+        await response.BufferContentAsync().ConfigureAwait(false);
+        return TryCreateErrorMessageFromResponse(response);
+    }
+
+    private static string TryCreateErrorMessageFromResponse(PipelineResponse response)
+    {
+        try
+        {
+            Internal.OpenAIError openAIError = Internal.OpenAIError.FromResponse(response);
+            return openAIError.ToExceptionMessage(response.Status);
+        }
+        catch (Exception)
```

Comment:
nit: IIRC it's preferred not to swallow base Exception types, as this could suppress Exceptions like OutOfMemoryException that should rightly bubble up to the end-user.  It's better to catch a specific exception type if possible.

Reply:
got it; I wasn't sure how conservative to be given we're already *trying* to throw an exception, but let me see if I can't account for that via `InnerException` instead of being overly gluttonous with the "..." catch behavior!

Reply:
It seemed sufficient to more narrowly constrain to `JsonException` during deserialization (which will handle unexpected input) and leave everything else (like OOMs/etc.) uncaught as "superseding" exceptions. Let me know if that seems insufficient; I'm sure we could do something fancy with ExceptionDispatchInfo and the like, but with the JsonException path covered I feel like anything else going wrong is *really* going wrong and novelly originated "exception during exception" wouldn't be inappropriate to convey bad system state at that point.